### PR TITLE
Improve private OCI registry for caas-image-repo

### PR DESCRIPTION
**Pull request by hpidcock**
- Ensure ecr private registry is used before docker registry. Until juju 3.3, we can't remove the default docker registry (i.e. using "jujusolutions" as a repo name instead of the fully qualified name "docker.io/jujusolutions"). This fixes an ordering issue with ecr private (and other repos where you can specify the repo without a subpath) being interpreted as a docker.io repos.
- Once you are using a private OCI registry for caas-image-repo, it is impossible to update the authentication details. This PR allows you to change those authentication details, but not change the registry/repository.

## QA steps

- Create private ECR registry
- Copy jujudb image + push jujud operator image
- Make sure to bootstrap with json repo info like:
- `juju bootstrap k8s --config caas-image-repo='{"repository":"12345.dkr.ecr.eu-west-1.amazonaws.com", "username":"AWS_KEY_ID", "password": "AWS_SECRET_KEY", "region":"eu-west-1" }'`
- Try updating the username and password:
- `juju controller-config caas-image-repo='{"repository":"12345.dkr.ecr.eu-west-1.amazonaws.com", "username":"NEW_AWS_KEY_ID_NEW", "password": "NEW_AWS_SECRET_KEY_NEW", "region":"eu-west-1" }'`
- Check image pull secrets are updated.
- Also try bootstrapping with invalid private ecr `juju bootstrap k8s --config caas-image-repo=12345.dkr.ecr.eu-west-1.amazonaws.com`. It should error `ERROR constructing controller config: empty credential for elastic container registry`.

## Documentation changes

N/A

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2037744

**Jira card:** JUJU-4707

### CODE REVIEW

**Review by ycliuhw — APPROVED** [review]
LGTM ,thanks for the fix!

File: cmd/juju/commands/bootstrap.go [hunk 1]
```diff
@@ -1497,6 +1498,21 @@ func (c *bootstrapCommand) bootstrapConfigs(
 		return bootstrapConfigs{}, errors.Annotate(err, "constructing bootstrap config")
 	}
 
+	// Pre-process controller attributes.
+	if _, ok := controllerConfigAttrs[controller.CAASOperatorImagePath]; ok {
+		return bootstrapConfigs{}, fmt.Errorf("%q is no longer supported controller configuration",
```
Comment (tlm) [inline cmd/juju/commands/bootstrap.go:1503]: Should we provide a reference to watch is in place now or some more documentation.

File: docker/registry/internal/acr.go [hunk 2]
```diff
@@ -29,6 +29,10 @@ func normalizeRepoDetailsAzure(repoDetails *docker.ImageRepoDetails) {
 	}
 }
 
+func (c *azureContainerRegistry) String() string {
```
Comment (tlm) [inline docker/registry/internal/acr.go:32]: I might be confused but the stringer interface just returns the domain and no other information?

File: docker/registry/internal/ecr.go [hunk 3]
```diff
@@ -146,13 +149,23 @@ func (c *elasticContainerRegistry) elasticContainerRegistryTransport(
...
 	}
-	if err := c.refreshTokenForElasticContainerRegistry(repoDetails); err != nil {
-		return nil, errors.Trace(err)
+	if repoDetails.Region == "" {
+		return nil, errors.NewNotValid(nil, "region is required")
 	}
-	if repoDetails.Auth.Empty() {
-		return nil, errors.NewNotValid(nil, "empty identity token for elastic container registry")
+	if repoDetails.Username == "" || repoDetails.Password == "" {
+		return nil, errors.NewNotValid(nil,
```
Comment (tlm) [inline docker/registry/internal/ecr.go:156]: NotValidf()?

File: docker/registry/internal/ecr.go [hunk 4]
```diff
@@ -146,13 +149,23 @@ func (c *elasticContainerRegistry) elasticContainerRegistryTransport(
...
+			fmt.Sprintf("username and password are required for registry %q", repoDetails.Repository),
+		)
 	}
-	return newBasicTransport(transport, "", "", repoDetails.Auth.Value), nil
+	return dynamicTransportFunc(func() (http.RoundTripper, error) {
+		if err := c.refreshTokenForElasticContainerRegistry(repoDetails); err != nil {
+			return nil, errors.Trace(err)
+		}
+		if repoDetails.Auth.Empty() {
+			return nil, errors.NewNotValid(nil, "empty identity token for elastic container registry")
```
Comment (tlm) [inline docker/registry/internal/ecr.go:165]: same here

File: make_functions.sh [hunk 5]
```diff
@@ -149,6 +150,23 @@ build_push_operator_image() {
...
 }
 
+seed_repository() {
+  set -x
+  docker pull "docker.io/jujusolutions/juju-db:${JUJU_DB_VERSION}"
+	docker tag "docker.io/jujusolutions/juju-db:${JUJU_DB_VERSION}" "${DOCKER_USERNAME}/juju-db:${JUJU_DB_VERSION}"
+	docker push "${DOCKER_USERNAME}/juju-db:${JUJU_DB_VERSION}"
+
+  # copy all the lts that are available
+  for (( i = 18; ; i += 2 )); do
```
Comment (tlm) [inline make_functions.sh:160]: Wowza

**Review by tlm — COMMENTED** [review]
All in all I am pretty happy with this. I would prefer the docker package to get renamed as it's not docker specific.

File: docker/registry/internal/acr.go (hunk 2 above)
Comment (hpidcock) [reply docker/registry/internal/acr.go:32]: Yep, its just to identify the repository ATM for tests.

File: docker/registry/internal/ecr.go (hunk 3 above)
Comment (hpidcock) [reply docker/registry/internal/ecr.go:156]: NotValidf has the wrong message.

File: make_functions.sh (hunk 5 above)
Comment (hpidcock) [reply make_functions.sh:160]: Look, it does what is says on the can. 🥫

File: cmd/juju/commands/bootstrap.go (hunk 1 above)
Comment (hpidcock) [reply cmd/juju/commands/bootstrap.go:1503]: I've got a jira ticket to update docs for caas-image-repo.

**Review by tlm — APPROVED** [review]
lgtm

**Pull request merged**